File: api/request-guard.ts
import { HttpStatus } from '../src/http-status.js';

interface GuardRequest {
  method?: string;
  headers: Record<string, string | string[] | undefined>;
}

export interface GuardFailure { status: number; code: string; retryAfter?: number }

const WINDOW_MS = 60_000;
const MAX_REQUESTS = 60;
const MAX_TRACKED_CLIENTS = 10_000;
const MAX_REQUEST_BYTES = 1_010_000;
const buckets = new Map<string, { count: number; resetsAt: number }>();

const firstHeader = (value: string | string[] | undefined): string | null =>
  typeof value === 'string' ? value : value?.[0] ?? null;

function clientKey(request: GuardRequest): string {
  const forwarded = firstHeader(request.headers['x-vercel-forwarded-for'])
    || firstHeader(request.headers['x-forwarded-for']);
  return forwarded?.split(',')[0]?.trim().slice(0, 100) || 'unknown';
}

function rateLimit(request: GuardRequest, now: number): GuardFailure | null {
  const key = clientKey(request);
  const current = buckets.get(key);
  if (!current || current.resetsAt <= now) {
    if (buckets.size >= MAX_TRACKED_CLIENTS) {
      for (const [candidate, bucket] of buckets) if (bucket.resetsAt <= now) buckets.delete(candidate);
      if (buckets.size >= MAX_TRACKED_CLIENTS) buckets.delete(buckets.keys().next().value as string);
    }
    buckets.set(key, { count: 1, resetsAt: now + WINDOW_MS });
    return null;
  }
  current.count += 1;
  if (current.count <= MAX_REQUESTS) return null;
  return { status: HttpStatus.TOO_MANY_REQUESTS, code: 'rate_limited', retryAfter: Math.ceil((current.resetsAt - now) / 1000) };
}

export function guardSnapshotRequest(request: GuardRequest, now = Date.now()): GuardFailure | null {
  const rateFailure = rateLimit(request, now);
  if (rateFailure) return rateFailure;
  if (request.method !== 'PUT') return null;
  const contentType = firstHeader(request.headers['content-type']);
  if (!contentType?.toLowerCase().startsWith('application/json')) {
    return { status: HttpStatus.UNSUPPORTED_MEDIA_TYPE, code: 'json_content_type_required' };
  }
  const contentLength = Number(firstHeader(request.headers['content-length']));
  if (Number.isFinite(contentLength) && contentLength > MAX_REQUEST_BYTES) {
    return { status: HttpStatus.CONTENT_TOO_LARGE, code: 'snapshot_too_large' };
  }
  return null;
}

export function resetRequestGuardForTests(): void { buckets.clear(); }

File: server/request_guard.py
from __future__ import annotations

from dataclasses import dataclass
from threading import Lock
from time import monotonic

MAX_REQUEST_BYTES = 1_010_000
MAX_REQUESTS = 60
WINDOW_SECONDS = 60
MAX_TRACKED_CLIENTS = 10_000


@dataclass(frozen=True, slots=True)
class GuardFailure:
    status: int
    code: str
    retry_after: int | None = None


class SnapshotRequestGuard:
    def __init__(self) -> None:
        self._buckets: dict[str, tuple[int, float]] = {}
        self._lock = Lock()

    def check(
        self,
        *,
        method: str,
        client_key: str,
        content_type: str | None,
        content_length: str | None,
        now: float | None = None,
    ) -> GuardFailure | None:
        current_time = monotonic() if now is None else now
        with self._lock:
            count, resets_at = self._buckets.get(client_key[:100], (0, current_time + WINDOW_SECONDS))
            if resets_at <= current_time:
                count, resets_at = 0, current_time + WINDOW_SECONDS
            count += 1
            self._buckets[client_key[:100]] = (count, resets_at)
            if len(self._buckets) > MAX_TRACKED_CLIENTS:
                expired = [key for key, (_, reset) in self._buckets.items() if reset <= current_time]
                for key in expired:
                    self._buckets.pop(key, None)
                while len(self._buckets) > MAX_TRACKED_CLIENTS:
                    self._buckets.pop(next(iter(self._buckets)))
            if count > MAX_REQUESTS:
                return GuardFailure(429, "rate_limited", max(1, int(resets_at - current_time)))

        if method != "PUT":
            return None
        if not (content_type or "").lower().startswith("application/json"):
            return GuardFailure(415, "json_content_type_required")
        try:
            declared_size = int(content_length or "0")
        except ValueError:
            declared_size = 0
        if declared_size > MAX_REQUEST_BYTES:
            return GuardFailure(413, "snapshot_too_large")
        return None

    def rate_limit_only(self, *, client_key: str, now: float | None = None) -> GuardFailure | None:
        """Routes that validate their own bodies still want the shared rate limit.

        Passing method="GET" to check() achieved this by accident and read as though
        the request really were a GET; this says what is meant.
        """
        return self.check(
            method="GET", client_key=client_key, content_type=None, content_length=None, now=now,
        )

    def reset(self) -> None:
        with self._lock:
            self._buckets.clear()
